_blog/2024-01-23-using-serde-rust.mdx

```diff
@@ -0,0 +1,341 @@
+
+```rust
+use serde::ser::{Serialize, Serializer, SerializeStruct};
+
+struct Color {
+    r: u8,
+    g: u8,
+    b: u8,
+}
+
+impl Serialize for Color {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // 3 is the number of fields in the struct.
+        let mut state = serializer.serialize_struct("Color", 3)?;
+        state.serialize_field("r", &self.r)?;
+        state.serialize_field("g", &self.g)?;
+        state.serialize_field("b", &self.b)?;
+        state.end()
+    }
+}
+```
+Note that to serialize a field, the field type also needs to implement `Serialize`. If you have a custom type that doesn't implement `Serialize`, you will either need to implement `Serialize` or use a `Serialize` derive macro (if the struct/enum type holds types that all implement `Serialize`).
+
+The `Deserialize` trait is a little bit different and is a fair bit more complicated to implement. To be able to deserialize to a type, the type itself needs to implement `Sized` which means that there are a number of types which can’t use this trait (for example `&str`) because they are unsized types. To deserialize a type, you also need to use a type that implements the `Visitor` trait. The `serde` documentation provides the below example for an `I32Visitor` type that attempts to deserialize multiple types to `i32`:
+
+```rust
+use std::fmt;
+
+use serde::de::{self, Visitor};
+
+struct I32Visitor;
+
+impl<'de> Visitor<'de> for I32Visitor {
+    type Value = i32;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("an integer between -2^31 and 2^31")
+    }
+
+    fn visit_i8<E>(self, value: i8) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(i32::from(value))
+    }
+
+    fn visit_i32<E>(self, value: i32) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(value)
+    }
+
+    fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        use std::i32;
+        if value >= i64::from(i32::MIN) && value <= i64::from(i32::MAX) {
+            Ok(value as i32)
+        } else {
+            Err(E::custom(format!("i32 out of range: {}", value)))
+        }
+    }
+
+    // Similar for other methods:
+    //   - visit_i16
+    //   - visit_u8
+    //   - visit_u16
+    //   - visit_u32
+    //   - visit_u64
+}
+```
+
+As you can see, the implementation is quite large! However, it also allows us to make the implementation much simpler. By implementing the `Visitor` trait, we can pass the Visitor type to our `Deserialize` method:
+
+```rust
+use serde::{Deserialize, Deserializer};
+
+impl<'de> Deserialize<'de> for i32 {
```

Comment:
Won't this be an issue ? I would assume that an implementation of `Deserialize<'de> for i32` already exists and that we can't have orphaned instance.

Reply:
The example is primarily for illustration and is supposed to be intended as a follow on to the previous snippet. I added a disclaimer that this snippet is primarily for understanding